warc_metadata_sidecar.py

```diff
@@ -0,0 +1,266 @@
+            mime = match_format.find('mime')
+            self.mime = mime.text if mime is not None else None
+            self.puid = self.get_puid(match_format)
+
+
+def find_mime_and_puid(fido, payload):
+    """Find the mimetype and preservation identifier using fido and python-magic."""
+    # Using fido to find mimetype.
+    fido.identify_stream(payload)
+    # Using python-magic to find mimetype.
+    payload.seek(0)
+    magic_mime = magic.from_buffer(payload.read(), mime=True)
+    mime_dict = {}
+    if fido.mime:
+        mime_dict['fido'] = fido.mime
+    if magic_mime:
+        mime_dict['python-magic'] = magic_mime
+    return mime_dict
+
+
+def find_character_set(bytes_payload):
+    """Find the character set of the payload using chardet."""
+    result = chardet.detect(bytes_payload)
+    result_dict = {'encoding': result['encoding'],
+                   'confidence': result['confidence']
+                   }
+    return result_dict
+
+
+def find_language(bytes_load):
+    """Find the language of the payload using pycld2."""
+    is_reliable, bytes_found, details = cld2.detect(BAD_CHARS.sub('',
+                                                                  bytes_load.decode('utf-8',
+                                                                                    'replace')),
+                                                    bestEffort=True)
+    new_list = []
+    # 'details' seems to always return 3, if the language is 'Unknown' we don't need to list it.
+    for item in details:
+        if item[0] != 'Unknown':
+            new_list.append({
+                'name': item[0],
+                'code': item[1],
+                'text-covered': item[2],
+                'score': item[3]})
+    if new_list:
+        lang_cld = {'reliable': is_reliable,
+                    'text-bytes': bytes_found,
+                    'languages': new_list}
+        return lang_cld
+    else:
+        return None
+
+
+def create_warcinfo_payload(new_file, operator=None, publisher=None):
+    """Collect WARC fields to create warcinfo record payload."""
+    hostname = socket.gethostname()
+    version = pkg_resources.require('warc-metadata-sidecar')[0].version
+    warc_doc = 'http://bibnum.bnf.fr/WARC/WARC_ISO_28500_version1_latestdraft.pdf'
+    warcinfo_payload = {'software': 'warc-metadata-sidecar/' + version,
+                        'hostname': hostname,
+                        'ip': socket.gethostbyname(hostname),
+                        'conformsTo': warc_doc,
+                        'description': 'WARC metdata sidecar for ' + new_file}
+    if publisher:
+        warcinfo_payload['publisher'] = publisher
+    if operator:
+        warcinfo_payload['operator'] = operator
+
+    return warcinfo_payload
+
+
+def create_string_payload(fido, mime_dict, result_dict, lang_cld):
+    puid = '\n{0} {1}'.format(PUID_TITLE, fido.puid) if fido.puid is not None else None
+    mime_str = '{0} {1}'.format(MIME_TITLE, mime_dict)
```

Comment:
If there is no mime then we would skip the record. So it should not be empty and it should not reach this function if it is empty.

Reply:
I thought Fido could determine a puid in some cases without finding a mimetype (or was that only if using extension checking? or maybe I am misremembering...). If so, I thought we should write the puid to a record.

Reply:
So do we want to record a puid even if we don't find a mime type? If we don't have a mime type it will not go through the process of getting the character set and language detection.